src/mono/mono/eventpipe/eventpipe-rt-mono.h

```diff
@@ -0,0 +1,949 @@
+// Implementation of eventpipe-rt.h targeting Mono runtime.
+#ifndef __EVENTPIPE_RT_MONO_H__
+#define __EVENTPIPE_RT_MONO_H__
+
+#include "eventpipe-rt-config.h"
+
+#ifdef ENABLE_PERFTRACING
+#include <config.h>
+#include <glib.h>
+#include <mono/utils/checked-build.h>
+#include <mono/utils/mono-coop-mutex.h>
+#include <mono/metadata/w32file.h>
+#include "eventpipe.h"
+
+#undef EP_ARRAY_SIZE
+#define EP_ARRAY_SIZE(expr) G_N_ELEMENTS(expr)
+
+#undef EP_INFINITE_WAIT
+#define EP_INFINITE_WAIT MONO_INFINITE_WAIT
+
+//TODO: Should make sure block is executed in safe mode.
+#undef EP_GCX_PREEMP_ENTER
+#define EP_GCX_PREEMP_ENTER {
```

Comment:
I'm not sure that these will become "GC Safe mode" on Mono.  The problem is that a lot of your locks are coop mutexes - which means that blocking calls on them already do a transition to GC Safe mode.  And we disallow nested transitions into GC Safe mode (this helps to catch logic errors).

So for example `ep_disable` transitioning to preemptive mode and then calling `disable` which tries to take the config lock would be a problem.

The general strategy in Mono is to keep most runtime code in GC Unsafe (cooperative) mode, and only transition to GC Safe (preemptive) mode narrowly around blocking operations.

I wonder if these locks need to be coop mutexes?

In general, a lock should be a coop mutex if: it will be held across a region that may include GC safepoints (for example allocation of managed objects or other GC Unsafe -> GC Safe transitions).  If the lock is held over a region where it is guaranteed that there will not be a GC safepoint, you can use a Mono OS mutex.  See more discussion here: https://github.com/mono/mono/pull/18860#issuecomment-586427783

Alternately, it may be fine to leave `EP_GCX_PREEMP_ENTER`/`EXIT` as no-ops on Mono, but we may need to add some GC Safe transitions on the genuinely potentially blocking operations (presumably I/O.  I haven't reached that code in reviewing yet.)

Reply:
Mainly just put them there for coreclr at the moment since they do this transition in their code. I believe they will be no-ops.

Reply:
I have not done any additional transitions in the code yet and the idea was to maybe switch to safe mode as part of entering icall, but if we prefer to be running as much as possible in unsafe then we will need to add some transitions into the code.

Reply:
Most code that could end up transition into safe is now called through the runtime shim and ends up in mono functions that already take care of doing the transition between unsafe/safe.